File: БР1.2/Соболев Артём/labs/lab1/FitnessAPI/src/main/java/io/github/artsobol/fitnessapi/feature/user/mapper/UserMapper.java
package io.github.artsobol.fitnessapi.feature.user.mapper;

import io.github.artsobol.fitnessapi.config.persistence.MapStructConfig;
import io.github.artsobol.fitnessapi.feature.user.dto.response.UserResponse;
import io.github.artsobol.fitnessapi.feature.user.entity.User;
import org.mapstruct.BeanMapping;
import org.mapstruct.Mapper;
import org.mapstruct.Mapping;

@Mapper(config = MapStructConfig.class)
public interface UserMapper {

    @BeanMapping(ignoreByDefault = true)
    @Mapping(target = "id")
    @Mapping(target = "username")
    @Mapping(target = "role")
    UserResponse toResponse(User user);
}

File: БР1.2/Алексеев Андрей/labs/lab1/src/main/kotlin/ru/itmo/pxdkxvan/lab1/user/mapper/UserMapper.kt
package ru.itmo.pxdkxvan.lab1.user.mapper

import org.mapstruct.AfterMapping
import org.mapstruct.Mapper
import org.mapstruct.Mapping
import org.mapstruct.MappingConstants
import org.mapstruct.MappingTarget
import ru.itmo.pxdkxvan.lab1.role.dto.RoleResponse
import ru.itmo.pxdkxvan.lab1.role.entity.RoleEntity
import ru.itmo.pxdkxvan.lab1.role.mapper.RoleMapper
import ru.itmo.pxdkxvan.lab1.user.dto.UserResponse
import ru.itmo.pxdkxvan.lab1.user.entity.UserAccountEntity

@Mapper(
    componentModel = MappingConstants.ComponentModel.SPRING,
    uses = [RoleMapper::class],
)
interface UserMapper {
    @Mapping(target = "isVerified", source = "verified")
    @Mapping(target = "roles", expression = "java(java.util.List.of())")
    fun toUserResponse(entity: UserAccountEntity): UserResponse

    @Mapping(target = "id", ignore = true)
    @Mapping(target = "createdAt", ignore = true)
    @Mapping(target = "updatedAt", ignore = true)
    @Mapping(target = "verified", ignore = true)
    fun fromRawData(
        firstName: String,
        lastName: String,
        middleName: String?,
        email: String,
        phone: String,
        passwordHash: String,
        roles: Collection<RoleEntity>,
    ): UserAccountEntity

    @Mapping(target = "isSystem", source = "system")
    fun toRoleResponse(entity: RoleEntity): RoleResponse

    @AfterMapping
    fun enrichUserResponse(entity: UserAccountEntity, @MappingTarget response: UserResponse): UserResponse =
        response.copy(
            roles = entity.roles.sortedBy { it.name }.map(this::toRoleResponse),
        )
}
